0005-private-communities.md

```diff
@@ -0,0 +1,87 @@
+- Feature Name: Private Communities
+- Start Date: 2024-02-15
+- RFC PR: [LemmyNet/rfcs#0005](https://github.com/LemmyNet/rfcs/pull/5)
+- Lemmy Issue: [LemmyNet/lemmy#187](https://github.com/LemmyNet/lemmy/issues/187)
+
+# Summary
+
+Lemmy is currently limited to public communities with public content. This RFC proposes a way to implement private, federated communities where only approved users can read and write.
+
+# Motivation
+
+Private spaces are a fundamental aspect of human communication. They allow discussing sensitive topics without interruption from outsiders. Supporting private communities in Lemmy gives a major new use case for talking among friends or within organizations. This use case is currently not covered by major Fediverse projects.
+
+# Detailed design
+
+The functionality requires only minor changes to the existing community implementation. The community profile with sidebar, icon etc remains public, so that anyone can follow the community. When a follow request is received, it needs to be manually approved by a moderator, using an interface similar to the existing registration applications. Posts and comments inside a private community are only visible to approved followers.
+
+This change is in some ways similar to [Local-only communities](https://github.com/LemmyNet/lemmy/pull/4350). That PR can be used as a reference for parts of code which need to be changed.
+
+## Database
+
+The `CommunityVisibility` enum used in `community.visibility` gets a new variant `Private`. Queries in `post_view.rs` and `comment_view.rs` need to be adjusted so that content in communities set to private is only shown to followers. This also needs to be covered by unit tests.
+
+There needs to be a new table to store follow requests:
+```sql
+CREATE TABLE community_follow_request (
```

Comment:
Instead of adding this table, we should just add a `accepted_application` to the `community_follower` table, or just not add any columns at all, and rely on the `SubscribedType::pending` for unapproved apps to private communities.

And I'd recommend that the new table is called `community_application` (this one's my fav), `private_community_application`, `community_registration_application` or something like that, and look almost like our existing `registration_application` table, with these columns:

- Answer, person_id, moderator_id (optional until acted upon), deny_reason (optional string)